### PR TITLE
Add support for node affinity to VM.Spec

### PR DESCRIPTION
**Pull request by MarSik**
This adds a new field to Vm.Spec.NodeAffinity that is directly
translated to pod's Pod.Spec.Affinity.NodeAffinity.

Fixes: #438

Signed-off-by: Martin Sivak <msivak@redhat.com>

### CODE REVIEW

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -78,6 +78,12 @@ func (t *templateService) RenderLaunchManifest(vm *v1.VM) (*kubev1.Pod, error) {
 		},
 	}
 
+	if vm.Spec.NodeAffinity != nil {
```
Comment (rmohr) [inline pkg/virt-controller/services/template.go]: We also add an affinity to the underlying pod, if we do a migration. Could you add a test, to make sure, that when rendering the migration Pod, we don't override such a pre-set affinity?

File: pkg/api/v1/types.go [hunk 2]
```diff
@@ -124,6 +124,8 @@ type VMSpec struct {
 	Domain *DomainSpec `json:"domain,omitempty"`
 	// If labels are specified, only nodes marked with all of these labels are considered when scheduling the VM.
 	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+	// Host affinity support
+	NodeAffinity *k8sv1.NodeAffinity `json:"nodeAffinity,omitempty"`
```
Comment (rmohr) [inline pkg/api/v1/types.go]: Could you add an Affinity meta struct where this is embeded? I think we already know, that the other affinities (e.g. PodAffinity) will join this one.

**Comment by michalskrivanek** [comment]
When there is a clear equivalent to this functionality at kubernetes level with POD affinity I believe we should _not_ add it to VM.spec. 
Are there other benefits?

**Comment by MarSik** [comment]
It has to get to the Pod somehow. So where else should the definition live?

**Comment by michalskrivanek** [comment]
ugh, right right, i missed the "directly translated" part, hehe:)
Great, are you thinking about how to use this feature as an end user?

**Comment by rmohr** [comment]
ok to test

**Comment by rmohr** [comment]
travis show sto failing unit tests

**Comment by rmohr** [comment]
retest this please

**Comment by rmohr** [comment]
Looks good so far. Now only a functional test is missing.

**Comment by fabiand** [comment]
@michalskrivanek as a user I know that certain servers have beefier disks, thus I assign some taint to the node, and the the correct toleration to the pod, to ensure that the pod will land on the beefy host.

**Comment by rmohr** [comment]
retest this please

File: glide.lock [hunk 3]
```diff
@@ -216,7 +216,7 @@ imports:
   - storage/v1
   - storage/v1beta1
 - name: k8s.io/apimachinery
-  version: 21de1bffe600f61f4873ae774a60f7a7d37ac9bf
+  version: 8a1a257c3a3503c77f25e5802e96e89a2a11ad61
```
Comment (rmohr) [inline glide.lock]: I think that this file does not need to be updated.

File: tests/vm_migration_test.go [hunk 4]
```diff
@@ -89,6 +89,52 @@ var _ = Describe("VmMigration", func() {
...
+			migration := tests.NewRandomMigrationForVm(sourceVM)
+			err = virtClient.RestClient().Post().Resource("migrations").Namespace(tests.NamespaceTestDefault).Body(migration).Do().Error()
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() v1.MigrationPhase {
+				obj, err := virtClient.RestClient().Get().Resource("migrations").Namespace(tests.NamespaceTestDefault).Name(migration.ObjectMeta.Name).Do().Get()
+				Expect(err).ToNot(HaveOccurred())
+				var m *v1.Migration = obj.(*v1.Migration)
+				return m.Status.Phase
+			}, TIMEOUT, POLLING_INTERVAL).Should(Equal(v1.MigrationSucceeded))
```
Comment (rmohr) [inline tests/vm_migration_test.go]: The test log shows that migration was not done after the timeout. The test looks like this should work. Did you try it yourself already?

File: tests/vm_migration_test.go [hunk 5]
```diff
@@ -89,6 +89,52 @@ var _ = Describe("VmMigration", func() {
...
+			vm, err := virtClient.RestClient().Post().Resource("vms").Namespace(tests.NamespaceTestDefault).Body(sourceVM).Do().Get()
+			Expect(err).ToNot(HaveOccurred())
+			tests.WaitForSuccessfulVMStart(vm)
+
+			migration := tests.NewRandomMigrationForVm(sourceVM)
+			err = virtClient.RestClient().Post().Resource("migrations").Namespace(tests.NamespaceTestDefault).Body(migration).Do().Error()
+			Expect(err).ToNot(HaveOccurred())
+
+			Eventually(func() v1.MigrationPhase {
+				obj, err := virtClient.RestClient().Get().Resource("migrations").Namespace(tests.NamespaceTestDefault).Name(migration.ObjectMeta.Name).Do().Get()
```
Comment (rmohr) [inline tests/vm_migration_test.go]: You can use the more convenient virtClient for such calls now:

```
virtClient.Migration(tests.NamespaceTestDefault).Get(name, k8smetav1.GetOptions{})
virtClient.Migration(tests.NamespaceTestDefault).Create(migration)
...
```

File: tests/vm_migration_test.go [hunk 6]
```diff
@@ -89,6 +89,52 @@ var _ = Describe("VmMigration", func() {
...
+				return m.Status.Phase
+			}, TIMEOUT, POLLING_INTERVAL).Should(Equal(v1.MigrationSucceeded))
+
+			// Check Pod and VM affinity
+			labelSelector, err := labels.Parse(v1.DomainLabel + "=" + sourceVM.ObjectMeta.Name + "," + v1.MigrationLabel + "=" + migration.ObjectMeta.Name)
+			Expect(err).ToNot(HaveOccurred())
+
+			pods, err := virtClient.CoreV1().Pods(tests.NamespaceTestDefault).List(metav1.ListOptions{LabelSelector: labelSelector.String()})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(pods.Items[0].Spec.Affinity.NodeAffinity).To(BeEquivalentTo(sourceVM.Spec.Affinity.NodeAffinity))
```
Comment (rmohr) [inline tests/vm_migration_test.go]: Shouldn't this check for the opposite? The NodeAffinity on the pod should be enriched with the anti-affinity based on the migration. Only the VM should still have the same affinity like it had before the migration.

**Comment by rmohr** [comment]
Follow up from an irc discussion: It turns out that the merging of the migration affinity rules needs to be done differently. Instead of adding another Term, we need to add the migration affinity rules to every pre-existing affinity rule as another requirement.

**Comment by rmohr** [comment]
retest this please

File: tests/vm_migration_test.go [hunk 7]
```diff
@@ -89,6 +89,52 @@ var _ = Describe("VmMigration", func() {
...
+			// Check Pod and VM affinity
+			labelSelector, err := labels.Parse(v1.DomainLabel + "=" + sourceVM.ObjectMeta.Name + "," + v1.MigrationLabel + "=" + migration.ObjectMeta.Name)
+			Expect(err).ToNot(HaveOccurred())
+
+			pods, err := virtClient.CoreV1().Pods(tests.NamespaceTestDefault).List(metav1.ListOptions{LabelSelector: labelSelector.String()})
+			Expect(err).ToNot(HaveOccurred())
+			Expect(pods.Items[0].Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions[0]).To(BeEquivalentTo(sourceVM.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms[0].MatchExpressions[0]))
+
+			close(done)
+		}, 30)
```
Comment (rmohr) [inline tests/vm_migration_test.go]: I would give it 90 seconds.

File: tests/vm_migration_test.go (hunk 4 above)
Comment (MarSik) [reply tests/vm_migration_test.go]: It timeouts for me locally and it worked in CI when I increased the global timeout to 30s

**Comment by rmohr** [comment]
@MarSik thx. Nice work. Do you mind having a look at https://github.com/kubevirt/user-guide and add a basic description? Something like linking to the kubernetes affinity page and describing the implact.

**Pull request merged**